lib/full-site-editing/block-templates.php

```diff
@@ -0,0 +1,281 @@
+	$template->theme       = $theme;
+	$template->content     = $post->post_content;
+	$template->slug        = $post->post_name;
+	$template->is_custom   = true;
+	$template->type        = $post->post_type;
+	$template->description = $post->post_excerpt;
+	$template->title       = $post->post_title;
+	$template->status      = $post->post_status;
+
+	return $template;
+}
+
+/**
+ * Retrieves a list of unified template objects based on a query.
+ *
+ * @param array $query Query.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return array Templates.
+ */
+function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_template' ) {
+	$wp_query_args = array(
+		'post_status'    => array( 'auto-draft', 'draft', 'publish' ),
+		'post_type'      => $template_type,
+		'posts_per_page' => -1,
+		'no_found_rows'  => true,
+	);
+
+	if ( isset( $query['theme'] ) ) {
+		$wp_query_args['tax_query'] = array(
+			array(
+				'taxonomy' => 'wp_theme',
+				'field'    => 'slug',
+				'terms'    => $query['theme'],
+			),
+		);
+	}
+
+	if ( isset( $query['slug__in'] ) ) {
+		$wp_query_args['post_name__in'] = $query['slug__in'];
+	}
+
+	// This is only needed for the regular templates/template parts CPT listing and editor.
+	if ( isset( $query['wp_id'] ) ) {
+		$wp_query_args['p'] = $query['wp_id'];
+	} else {
+		$wp_query_args['post_status'] = 'publish';
+	}
+	$template_query = new WP_Query( $wp_query_args );
+	$query_result   = array();
+	foreach ( $template_query->get_posts() as $post ) {
+		$query_result[] = _gutenberg_build_template_result_from_post( $post );
+	}
+
+	if ( ! isset( $query['wp_id'] ) && ( ! isset( $query['theme'] ) || wp_get_theme()->get_stylesheet() === $query['theme'] ) ) {
+		$template_files = _gutenberg_get_template_files( $template_type );
+		foreach ( $template_files as $template_file ) {
+			$is_custom      = array_search(
+				wp_get_theme()->get_stylesheet() . '/' . $template_file['slug'],
+				array_column( $query_result, 'id' ),
+				true
+			);
+			$should_include = false === $is_custom && (
+				! isset( $query['slug__in'] ) || in_array( $template_file['slug'], $query['slug__in'], true )
+			);
+			if ( $should_include ) {
+				$query_result[] = _gutenberg_build_template_result_from_file( $template_file, $template_type );
+			}
+		}
+	}
+
+	return $query_result;
+}
+
+/**
+ * Retrieves a single unified template object using its id.
+ *
+ * @param array $id Template id.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return WP_Block_Template|null Template.
+ */
+function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
+	$parts = explode( '|', $id, 2 );
+	if ( count( $parts ) < 2 ) {
+		return null;
+	}
+	list( $theme, $slug ) = $parts;
+	$wp_query_args        = array(
+		'name'           => $slug,
+		'post_type'      => $template_type,
+		'post_status'    => array( 'auto-draft', 'draft', 'publish', 'trash' ),
```

Comment:
Do we really want all these statuses? Even trashed ones? I'm thinking e.g. of a user that uses the wp-admin > Appearance > Templates screen to trash a template they no longer want :thinking: